xds/internal/xdsclient/bootstrap/bootstrap.go

```diff
@@ -141,12 +192,56 @@ func (sc *ServerConfig) UnmarshalJSON(data []byte) error {
 		if err != nil {
 			return fmt.Errorf("failed to build credentials bundle from bootstrap for %q: %v", cc.Type, err)
 		}
-		sc.Creds = grpc.WithCredentialsBundle(bundle)
+		sc.Creds = ChannelCreds{
+			Type:   cc.Type,
+			Config: cc.Config,
+		}
+		sc.credsDialOption = grpc.WithCredentialsBundle(bundle)
 		break
 	}
 	return nil
 }
 
+// ServerConfigFromJSON creates a new ServerConfig from the given JSON
+// configuration. This is the preferred way of creating a ServerConfig when
+// hand-crafting the JSON configuration.
+func ServerConfigFromJSON(data []byte) (*ServerConfig, error) {
+	sc := new(ServerConfig)
+	if err := sc.UnmarshalJSON(data); err != nil {
+		return nil, err
+	}
+	return sc, nil
+}
+
+// Equal reports whether sc and other are considered equal.
+func (sc *ServerConfig) Equal(other *ServerConfig) bool {
+	switch {
+	case sc == nil && other == nil:
+		return true
+	case (sc != nil) != (other != nil):
+		return false
```

Comment:
FWIW: `if sc == nil || other == nil` is equivalent given the preceding `case`; I'm not sure which is more readable, but I find this form with 3 `!=`s to be a little confusing, personally.

Reply:
I agree that `sc == nil || other == nil` is much easier to read than `(sc != nil) != (other != nil)` when considered in isolation. But when used here, the former requires carrying over of state from the previous `case`, because we need to remember that both are not `nil` to make complete sense of this statement. But the latter is more self-contained as it simply compares the `nil`ness of one with the other.

I think I might have written it as `if sc == nil || other == nil` if I had written this whole method as a series of `if` conditions, instead of a case statement.